File: node/smtpd.js
/*
 * Simple SMTP server in Node.JS
 *
 * Author: Maarten Oelering
 */
 
var net = require('net'),
    sys = require('util'),
    os = require('os');

var listenPort = 2525;
var listenAddress = "0.0.0.0";
var hostname = os.hostname();

var server = net.createServer(function (socket) {
  
  var buffer = '';
  var dataState = false;
  var content = '';
  
  socket.setEncoding('ascii');

  socket.on('connect', function () {
    socket.write('220 ' + hostname + ' ESMTP\r\n');
  });

  socket.on('data', function (data) {
    buffer += data;
    var start = 0;
    while (true) {
      // find next LF in buffer
      var i = buffer.indexOf('\n', start);
      if (i == -1) {
        // keep part after last LF in buffer
        if (start > 0)
          buffer = buffer.slice(start);
        break;
      }
      // trim trailing CR from line
      var end = (i > start && buffer.charAt(i - 1) == '\r') ? (i - 1) : i;
      // emit line without trailing (CR)LF
      socket.emit('line', buffer.substring(start, end));
      start = i + 1;
      // buffer ended with LF
      if (start == buffer.length) {
        buffer = '';
        break;
      }
    }
  });

  socket.on('line', function (line) {
    if (!dataState) {
      var command = line.substring(0, 4);
      switch (command.toUpperCase()) {
        case 'HELO':
          socket.write('250 ' + hostname + '\r\n');
          break;
        case 'EHLO':
          socket.write('250-' + hostname + '\r\n250 PIPELINING\r\n');
          break;
        case 'MAIL':
          socket.write('250 OK\r\n');
          break;
        case 'RCPT':
          socket.write('250 OK\r\n');
          break;
        case 'DATA':
          dataState = true;
          socket.write('354 End data with <CR><LF>.<CR><LF>\r\n');
          break;
        case 'RSET':
          socket.write('250 OK\r\n');
          break;
        case 'QUIT':
          // close connection after sending response
          socket.end('221 ' + hostname + ' closing connection\r\n');
          break;
        default:
          socket.write('500 unrecognized command\r\n');
          break;
      }
    }
    else {
      if (line == '.') {
        dataState = false;
        socket.write('250 OK\r\n');
      }
      else {
        // todo: unescape /^\./
        content += line;
        content += '\r\n';
      }
    }
  });
  
  socket.on('timeout', function () {
    console.log("timeout");
  });

  socket.on('end', function () {
    // do nothing
  });

  socket.on('error', function (e) {
    console.log("error:" + e);
  });
});

server.listen(listenPort, listenAddress);

File: go/smtpd.go
/* 
 * Simple SMTP server in Go
 *
 * Author: Maarten Oelering
 */
package main

import (
    "fmt"
    "os"
    "net"
    "bufio"
    "strings"
)

var hostname string;

func main() {
    hostname, _ = os.Hostname()

    tcpAddress, _ := net.ResolveTCPAddr(":2525")
    listener, _ := net.ListenTCP("tcp", tcpAddress)
    for {
        conn, _ := listener.AcceptTCP()
        go handleConnection(conn)
    }
    listener.Close()
}

func handleConnection(conn *net.TCPConn) {
    defer conn.Close()

    reader := bufio.NewReader(conn)
    
    fmt.Fprintf(conn, "220 %s ESMTP\r\n", hostname)
    
    dataMode := false
    for {
        line, _ := reader.ReadString('\n')
        if ! dataMode {
            command := strings.ToUpper(line[0:4])
            switch (command) {
            case "HELO":
                fmt.Fprintf(conn, "250 %s\r\n", hostname)
            case "EHLO":
                fmt.Fprintf(conn, "250-%s\r\n250 PIPELINING\r\n", hostname)
            case "MAIL":
                conn.Write([]byte("250 OK\r\n"))
            case "RCPT":
                conn.Write([]byte("250 OK\r\n"))
            case "DATA":
                dataMode = true
                conn.Write([]byte("354 End data with <CR><LF>.<CR><LF>\r\n"))
            case "QUIT":
                fmt.Fprintf(conn, "221 %s closing connection\r\n", hostname)
                return  // and close
            default:
                conn.Write([]byte("500 unrecognized command\r\n"))
            }
        } else {
            if len(line) >= 2 && line[0] == '.' && (line[1] == '\r' || line[1] == '\n') {
                dataMode = false
                conn.Write([]byte("250 OK\r\n"))
            } else {
                if strings.HasPrefix(line, "..") {
                    line = line[1:]
                }
                // TODO: collect data in var
            }
        }
    }
}
